% Template for PLoS
% Version 3.1 February 2015
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template
%
% % % % % % % % % % % % % % % % % % % % % %
%
% -- IMPORTANT NOTE
%
% This template contains comments intended 
% to minimize problems and delays during our production 
% process. Please follow the template instructions
% whenever possible.
%
% % % % % % % % % % % % % % % % % % % % % % % 
%
% Once your paper is accepted for publication, 
% PLEASE REMOVE ALL TRACKED CHANGES in this file and leave only
% the final text of your manuscript.
%
% There are no restrictions on package use within the LaTeX files except that 
% no packages listed in the template may be deleted.
%
% Please do not include colors or graphics in the text.
%
% Please do not create a heading level below \subsection. For 3rd level headings, use \paragraph{}.
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% -- FIGURES AND TABLES
%
% Please include tables/figure captions directly after the paragraph where they are first cited in the text.
%
% DO NOT INCLUDE GRAPHICS IN YOUR MANUSCRIPT
% - Figures should be uploaded separately from your manuscript file. 
% - Figures generated using LaTeX should be extracted and removed from the PDF before submission. 
% - Figures containing multiple panels/subfigures must be combined into one image file before submission.
% For figure citations, please use "Fig." instead of "Figure".
% See http://www.plosone.org/static/figureGuidelines for PLOS figure guidelines.
%
% Tables should be cell-based and may not contain:
% - tabs/spacing/line breaks within cells to alter layout or alignment
% - vertically-merged cells (no tabular environments within tabular environments, do not use \multirow)
% - colors, shading, or graphic objects
% See http://www.plosone.org/static/figureGuidelines#tables for table guidelines.
%
% For tables that exceed the width of the text column, use the adjustwidth environment as illustrated in the example table in text below.
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% -- EQUATIONS, MATH SYMBOLS, SUBSCRIPTS, AND SUPERSCRIPTS
%
% IMPORTANT
% Below are a few tips to help format your equations and other special characters according to our specifications. For more tips to help reduce the possibility of formatting errors during conversion, please see our LaTeX guidelines at http://www.plosone.org/static/latexGuidelines
%
% Please be sure to include all portions of an equation in the math environment.
%
% Do not include text that is not math in the math environment. For example, CO2 will be CO\textsubscript{2}.
%
% Please add line breaks to long display equations when possible in order to fit size of the column. 
%
% For inline equations, please do not include punctuation (commas, etc) within the math environment unless this is part of the equation.
%
% % % % % % % % % % % % % % % % % % % % % % % % 
%
% Please contact latex@plos.org with any questions.
%
% % % % % % % % % % % % % % % % % % % % % % % %

\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% fixltx2e package for \textsubscript
\usepackage{fixltx2e}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% rotating package for sideways tables
\usepackage{rotating}

% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Additional packages used by authors
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% To allow code "Boxes" as well as figures
\usepackage{fancyvrb}
\usepackage{newfloat}
\DeclareFloatingEnvironment[name={Box}]{codebox}
\usepackage{listings} % For code examples
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor}
\usepackage{verbatim}

% Do not hyphenate the following words
\hyphenation{MODELLER}

% For the \num command
\usepackage[group-separator={,}]{siunitx}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% End additional packages
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\sf PLOS}

%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}

%% END MACROS SECTION


\begin{document}
\vspace*{0.35in}

% Title must be 250 characters or less.
% Please capitalize all terms in the title except conjunctions, prepositions, and articles.
\begin{flushleft}
{\Large
\textbf\newline{Ensembler: Enabling High-Throughput Molecular Simulations at the Superfamily Scale}
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Daniel L. Parton\textsuperscript{1,},
Patrick B. Grinaway\textsuperscript{1},
Sonya M. Hansom\textsuperscript{1},
Kyle A. Beauchamp\textsuperscript{1},
John D. Chodera\textsuperscript{1,*},
\\
\bigskip
\bf{1} Computational Biology Program, Sloan Kettering Institute, Memorial Sloan Kettering Cancer Center, New York, NY 10065
\\
\bigskip

% Insert additional author notes using the symbols described below. Insert symbol callouts after author names as necessary.
% 
% Remove or comment out the author notes below if they aren't used.
%

% Use the asterisk to denote corresponding authorship and provide email address in note below.
* john.chodera@choderalab.org

\end{flushleft}
% Please keep the abstract below 300 words
\section*{Abstract}
The rapidly expanding body of available genomic and protein structural data provides a rich resource for understanding protein dynamics with biomolecular simulation. 
While computational infrastructure has grown rapidly, simulations on an \emph{omics} scale are not yet widespread, primarily because software infrastructure to enable simulations at this scale has not kept pace. 
It should now be possible to study protein dynamics across entire (super)families, exploiting both available structural biology data and conformational similarities across homologous proteins.
Here, we present a new tool for enabling high-throughput simulation in the genomics era.
{\bf Ensembler} takes any set of sequences---from a single sequence to an entire superfamily---and shepherds them through various stages of modeling and refinement to produce simulation-ready structures.
This includes comparative modeling to all relevant PDB structures (which may span multiple conformational states of interest), reconstruction of missing loops, addition of missing atoms, culling of nearly identical structures, assignment of appropriate protonation states, solvation in explicit solvent, and refinement and filtering with molecular simulation to ensure stable simulation. 
The output of this pipeline is an ensemble of structures ready for subsequent molecular simulations using computer clusters, supercomputers, or distributed computing projects like Folding@home.
{\bf Ensembler} thus automates much of the time-consuming process of preparing protein models suitable for simulation, while allowing scalability up to entire superfamilies.
A particular advantage of this approach can be found in the construction of kinetic models of conformational dynamics---such as Markov state models (MSMs)---which benefit from a diverse array of initial configurations that span the accessible conformational states to aid sampling.
We demonstrate the power of this approach by constructing models for all catalytic domains in the human tyrosine kinase family, using all available kinase catalytic domain structures from any organism as structural templates.

{\bf Ensembler} is free and open source software licensed under the GNU General Public License (GPL) v2. 
It is compatible with Linux and OS X.
The latest release can be installed via the {\tt conda} package manager, and the latest source can be downloaded from \url{https://github.com/choderalab/ensembler}.

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step. 
% Author Summary not valid for PLOS ONE submissions.   
\section*{Author Summary}
% TODO improve
Proteins are the workhorses of the human body, and are involved in essentially every biological process.
Many diseases are caused by proteins malfunctioning.
To understand how a protein functions, it is necessary to know its physical properties.
The field of structural biology provides many techniques for determining the three-dimensional structure of a protein.
The dynamics of a protein, i.e. the way it moves, are of equal importance, but are more difficult to uncover with traditional experimental techniques.
Computer simulations are an effective alternative method for understanding protein dynamics, but require experimental structural information as a starting point.
While recent advances in genomics and experimental techniques have provided a wealth of such structural data, the appropriate software for using this data effectively has been lacking.
To tackle this problem, we have developed a software package called {\bf Ensembler}, which allows a user to automatically select appropriate experimentally derived structures for a given protein or family of proteins, and to use them to prepare a series of simulations.
The resultant simulation data can then used to investigate the dynamics of the protein(s) in question, and their involvement in disease.

\linenumbers

\section*{Introduction}
Recent advances in genomics and structural biology have helped generate an enormous wealth of protein data at the level of amino-acid sequence and three-dimensional structure.
However, proteins typically exist as an ensemble of thermally accessible conformational states, and static structures provide only a snapshot of their rich dynamical behavior.
Many functional properties---such as the ability to bind small molecules or interact with signaling partners---require transitions between states, encompassing anything from reorganization of sidechains at binding interfaces to domain motions to large scale folding-unfolding events.
Drug discovery could also benefit from a more extensive consideration of protein dynamics, whereby small molecules might be selected based on their predicted ability to bind and trap a protein target in an inactive state~\cite{craik:science:2009:trapping-moving-targets}.

Molecular dynamics (MD) simulations have the capability, in principle, to describe the time evolution of a protein in atomistic detail, and have proven themselves to be a useful tool in the study of protein dynamics.
A number of mature software packages and forcefields are now available, and much recent progress has been driven by advances in computing architecture.
For example, many MD packages are now able to exploit GPUs~\cite{eastman:jctc:2012:openmm,walker:jctc:2013:amber-gpu}, which provide greatly improved simulation efficiency per unit cost relative to CPUs, while distributed computing platforms such as Folding@home~\cite{shirts-pande:science:2000:fah}, Copernicus~\cite{pronk:2011:copernicus,pronk:2015:copernicus}, and GPUGrid~\cite{buch:2010:gpugrid}, allow scalability on an unprecedented level.
In parallel, methods for building human-understandable models of protein dynamics from noisy simulation data, such as Markov state modeling (MSM) approaches, are now reaching maturity~\cite{pande-beauchamp-bowman:2010:methods:markov-model-review,noe:jcp:2011:msm-review,chodera-noe:curr-opin-struct-biol:2014:msm-review}.
MSM methods in particular have the advantage of being able to aggregate data from multiple independent MD trajectories, facilitating parallelization of production simulations and thus greatly alleviating overall computational cost.
There also exist a number of mature software packages for comparative modeling of protein structures, in which a target protein sequence is modeled using one or more structures as templates~\cite{moult:proteins:2014:casp10,baker:science:2001:structure-prediction-review,baker:structure:2013:rosettacm,yang:nature-methods:2015:i-tasser-suite}.


However, it remains difficult for researchers to exploit the full variety of available protein sequence data (in simulating groups of related proteins) and structural data (exploiting multiple structures for each protein and its homologs/orthologs) in simulation studies in molecular simulations, largely due to limitations in software architecture.
For example, the preparation of a biomolecular simulation is typically performed manually, encompassing a series of fairly standard (yet time-consuming) steps such as the choice of protein sequence construct and starting structure(s), addition of missing residues and atoms, solvation with explicit water and counterions (and potentially buffer components and cosolvents), choice of simulation parameters (or parameterization schemes for components where parameters do not yet exist),  system relaxation with energy minimization, and one or more short preparatory MD simulations to equilibrate the system and relax the simulation cell.
Due to the laborious and manual nature of this process, simulation studies typically consider only one or a few proteins and starting configurations,  though notable exceptions exist, such as the Dynameomics effort of Daggett and coworkers in which over 100 proteins have been simulated so far using a single initial configuration for each~\cite{daggett:structure:2010:dynameomics}.
Worse still, studies (or collections of studies) that \emph{do} consider multiple proteins often suffer from the lack of consistent best practices in this preparation process, making comparisons between related proteins unnecessarily difficult.

The ability to fully exploit the large quantity of available protein sequence and structural data in biomolecular simulation studies could open up many interesting avenues for research, enabling the study of entire protein families or superfamilies within a single organism or across multiple organisms.
The similarity between members of a given protein family could be exploited to generate arrays of conformational models for related sequences, which could be used as starting configurations to aid sampling in MD simulations.
The conformations captured in structures of related members has been shown to provide useful information about the conformations accessible to all members of the family~\cite{friedland:plos-comput-biol:2009:conformational-dynamics,sali:2012:allostery-modeling}, though energetic differences between individuals will modify the populations and dynamics of individual conformational states.
This approach would be highly beneficial for many MD methods, such as MSM construction, which require global coverage of the conformational landscape to realize their full potential, and would also be particularly useful in cases where structural data is present for only a subset of the members of a protein family.
It would also aid in studying protein families known to have multiple metastable conformations---such as kinases---for which the combined body of structural data for the family may cover a large range of these conformations, while the available structures for any individual member might encompass only one or two distinct conformations.

Here, we present the first steps toward bridging the gap between biomolecular simulation software and \emph{omics}-scale sequence and structural data: a fully automated open source framework for building simulation-ready protein models in multiple conformational substates scalable from single sequences to entire superfamilies.
{\bf Ensembler} provides functions for selecting target sequences and homologous template structures, and (by interfacing with a number of external packages) performs pairwise alignments, comparative modeling of target-template pairs, and several stages of model refinement.
As an example application, we have constructed models for the entire set of human tyrosine kinase (TK) catalytic domains, using all available structures of protein kinase domains (from any species) as templates.
This results in a total of almost 400,000 models, and we demonstrate that these provide wide-ranging coverage of known functionally relevant conformations.
By using these models as starting configurations for highly parallel MD simulations, we expect their structural diversity to greatly aid in sampling of conformational space.
We further suggest that models with high target-template sequence identity are the most likely to represent native metastable states, while lower sequence identity models would aid in sampling of more distant regions of accessible phase space.
It is also important to note that some models (especially low sequence identity models) may not represent natively accessible conformations.
However, MSM methods benefit from the ability to remove outlier MD trajectories which start from non-natively accessible conformations, and which would thus be unconnected with the phase space sampled in other trajectories.
These methods essentially identify the largest subset of Markov nodes which constitute an ergodic network~\cite{msmbuilder,caflisch:jpcb:2011:msm-subgraph}.

We anticipate that {\bf Ensembler} will prove to be useful in a number of other ways.
For example, the generated models could represent valuable data sets even without subsequent production simulation, allowing exploration of the conformational diversity present within the available structural data for a given protein family.
Furthermore, automation of simulation preparation provides an excellent opportunity to make concrete certain ``best practices'', such as the choice of simulation parameters, approach to the treatment of protonation states, treatment of cofactors and structural ions, and pre-simulation refinement and equilibration procedures.
While the current version of {\bf Ensembler} only codifies some of these choices as default parameters, its modular nature allows additional stages to be easily added in the future.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Design and Implementation
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Design and Implementation}

{\bf Ensembler} is written in Python, and can be used via a command-line tool ({\tt ensembler}) or via a flexible Python API to allow integration of its components into other applications.
All command-line and API information in this article refers to the \href{https://github.com/choderalab/ensembler/releases/tag/v1.0.6}{version 1.0.6 release of Ensembler}.
Up-to-date documentation can be found at \href{http://ensembler.readthedocs.org/en/latest/}{ensembler.readthedocs.org}.

The {\bf Ensembler} modeling pipeline comprises a series of stages which are performed in a defined order. 
A visual overview of the pipeline is shown in Fig.~\ref{figure:pipeline}.
The various stages of this pipeline are described in detail below.

\begin{figure}[h]
    % \includegraphics[width=1.0\textwidth]{pipeline/pipeline2}

  \caption{{\bf Diagrammatic representation of the stages of the Ensembler pipeline and illustrative statistics for modeling all human tyrosine kinase catalytic domains.}
  On the left, the various stages of the {\bf Ensembler} pipeline are shown.
  The red labels indicate the corresponding text description provided for each stage in the Design and Implementation section.
  On the right, the number of viable models surviving each stage of the pipeline is shown for the 93 target TK domains and for two representative individual TK domains (\emph{SRC} and \emph{ABL}).
  Typical timings on a computer cluster (containing Intel Xeon E5-2665 2.4GHz hyperthreaded processors and NVIDIA GTX-680 or GTX-Titan GPUs) is reported to illustrate resource requirements per model for modeling the entire set of tyrosine kinases.
  Note that \emph{CPU-h} denotes the number of hours consumed by the equivalent of a single CPU hyperthread and \emph{GPU-h} on a single GPU---parallel execution via MPI reduces wall clock time nearly linearly.
  }
  \label{figure:pipeline}
\end{figure}

\subsection*{Target selection and retrieval (A)}

The first stage entails the selection of a set of \emph{target} protein sequences---the sequences for which the user is interested in generating simulation-ready structural models.
This may be a single sequence---such as a full-length protein or a construct representing a single domain---or a collection of sequences, such as a particular domain from an entire family of proteins.
The output of this stage is a FASTA-formatted text file containing the desired target sequences with corresponding arbitrary identifiers.

The {\tt ensembler} command-line tool allows targets to be selected from UniProt---a freely accessible resource for protein sequence and functional data (\href{http://www.uniprot.org/}{uniprot.org})~\cite{uniprot:2015}---via a UniProt search query.
To retrieve target sequences from UniProt, the subcommand {\tt gather\_targets} is used with the {\tt -{}-query} flag followed by a UniProt query string conforming to the same syntax as the search function available on the UniProt website.
For example, {\tt -{}-query `mnemonic:SRC\_HUMAN'} would select the full-length human Src sequence, while the query shown in Box~\ref{box:tk-gather-targets-templates} would select all human tyrosine protein kinases which have been reviewed by a human curator.
In this way, the user may select a single protein, many proteins, or an entire superfamily from UniProt.
The program outputs a FASTA file, setting the UniProt mnemonic (e.g.~{\tt SRC\_HUMAN}) as the identifier for each target protein.

\begin{codebox}[h]
\begin{adjustwidth}{-2.25in}{0in}
\footnotesize
\begin{Verbatim}[frame=single]
ensembler gather_targets --query `family:"tyr protein kinase family" AND organism:"homo sapiens" AND reviewed:yes'
                         --uniprot_domain_regex `^Protein kinase(?!; truncated)(?!; inactive)'
ensembler gather_templates --gather_from uniprot --query `domain:"Protein kinase" AND reviewed:yes'
                           --uniprot_domain_regex `^Protein kinase(?!; truncated)(?!; inactive)'
\end{Verbatim}
\caption{{\bf Ensembler command-line functions used to select targets and templates.}
The commands retrieve target and template data by querying UniProt.
The query string provided to the {\tt gather\_targets} command selects all human tyrosine protein kinases which have been reviewed by a curator, while the query string provided to the {\tt gather\_templates} command selects all reviewed protein kinases of any species.
The {\tt-{}-uniprot\_domain\_regex} flag is used to select a subset of the domains belonging to the returned UniProt protein entries, by matching the domain annotations against a given regular expression.
In this example, domains of type `Protein kinase', `Protein kinase 1', and `Protein kinase 2' were selected, while excluding many other domain types such as `Protein kinase; truncated', `Protein kinase; inactive', `SH2', `SH3', etc.
Target selection simply entails the selection of sequences corresponding to each matching UniProt domain.
Template selection entails the selection of the sequences and structures of any PDB entries corresponding to the matching UniProt domains.
}
\label{box:tk-gather-targets-templates}
\end{adjustwidth}
\end{codebox}

In many cases, it will be desirable to build models of an isolated protein domain, rather than the full-length protein.
The {\tt gather\_targets} subcommand allows protein domains to be selected from UniProt data by passing a regular expression string to the {\tt -{}-uniprot\_domain\_regex} flag. 
For example, the above {\tt -{}-query} flag for selecting all human protein kinases returns UniProt entries with domain annotations including `Protein kinase', `Protein kinase 1', `Protein kinase 2', `Protein kinase; truncated', `Protein kinase; inactive', `SH2', `SH3', etc.
The regular expression shown in Box~\ref{box:tk-gather-targets-templates} selects only domains of the first three types.
If the {\tt -{}-uniprot\_domain\_regex} flag is used, target identifiers are set with the form {\tt [UniProt mnemonic]\_D[domain index]}, where the latter part represents a 0-based index for the domain---necessary because a single target protein may contain multiple domains of interest (e.g.~{\tt JAK1\_HUMAN\_D0}, {\tt JAK1\_HUMAN\_D1}).

Target sequences can also be defined manually (or from another program) by providing a FASTA-formatted text file containing the desired target sequences with corresponding arbitrary identifiers.

\subsection*{Template selection and retrieval (B)}

{\bf Ensembler} uses comparative modeling to build models, and as such requires a set of structures to be used as templates.
The second stage thus entails the selection of templates and storage of associated sequences, structures, and identifiers.
These templates can be specified manually, or using the {\tt ensembler gather\_templates} subcommand to automatically select templates based on a search of the Protein Data Bank (PDB) or UniProt.
A recommended approach is to select templates from UniProt which belong to the same protein family as the targets, establishing homology and some degree of sequence identity between targets and templates.

The {\tt ensembler gather\_templates} subcommand provides methods for selecting template structures from either UniProt or the PDB (\url{http://www.rcsb.org/pdb}), specified by the {\tt -{}-gather\_from} flag.
Both methods select templates at the level of PDB chains---a PDB structure containing multiple chains with identical sequence spans (e.g.~for crystals with non-crystallographic symmetry giving rise to independent conformations of the protein within the asymmetric unit) would thus give rise to multiple template structures.

Selection of templates from the PDB simply requires passing a list of PDB IDs as a comma-separated string, e.g.~{\tt -{}-query 2H8H,1Y57}.
Specific PDB chain IDs can optionally also be selected via the {\tt -{}-chainids} flag.
The program retrieves structures from the PDB server, as well as associated data from the SIFTS service (\href{http://www.ebi.ac.uk/pdbe/docs/sifts/}{www.ebi.ac.uk/pdbe/docs/sifts})~\cite{velankar:2013:sifts}, which provides residue-level mappings between PDB and UniProt entries.
The SIFTS data is used to extract template sequences, retaining only residues which are resolved and match the equivalent residue in the UniProt sequence---non-wildtype residues are thus removed from the template structures.
Furthermore, PDB chains with less than a given percentage of resolved residues (default: 70\%) are filtered out.
Sequences are stored in a FASTA file, with identifiers of the form {\tt [UniProt mnemonic]\_D[UniProt domain index]\_[PDB ID]\_[PDB chain ID]}, e.g. {\tt SRC\_HUMAN\_D0\_2H8H\_A}.
Matching residues then extracted from the original coordinate files and stored as PDB-format coordinate files.

Selection of templates from UniProt proceeds in a similar fashion as for target selection; the {\tt -{}-query} flag is used to select full-length proteins from UniProt, while the optional {\tt -{}-uniprot\_domain\_regex} flag allows selection of individual domains with a regular expression string (Box~\ref{box:tk-gather-targets-templates}).
The returned UniProt data for each protein includes a list of associated PDB chains and their residue spans, and this information is used to select template structures, using the same method as for template selection from the PDB.
Only structures solved by X-ray crystallography or NMR are selected, thus excluding computer-generated models available from the PDB.
If the {\tt -{}-uniprot\_domain\_regex} flag is used, then templates are truncated at the start and end of the domain sequence.

Templates can also be defined manually.
Manual specification of templates simply requires storing the sequences and arbitrary identifiers in a FASTA file, and the structures as PDB-format coordinate files with filenames matching the identifiers in the sequence file.
The structure residues must also match those in the sequence file.

\subsection*{Template refinement (C)}

Unresolved template residues can optionally be modeled into template structures with the {\tt loopmodel} subcommand, which employs a kinematic closure algorithm provided via the {\tt loopmodel} tool of the Rosetta software suite~\cite{qian:nature:2007:modeller,wang:jmb:2007:modeller}.
We expect that in certain cases, pre-building template loops with Rosetta {\tt loopmodel} prior to the main modeling stage (with MODELLER) may result in improved model quality.
Loop remodeling may fail for a small proportion of templates due to spatial constraints imposed by the original structure; the subsequent modeling step thus automatically uses the remodeled version of a template if available, but otherwise falls back to using the non-remodeled version.
Furthermore, the Rosetta {\tt loopmodel} program will not model missing residues at the termini of a structure---such residue spans are modeled in the subsequent stage.

\subsection*{Alignment and comparative modeling (D)}

In the modeling stage, structural models of the target sequence are generated from the template structures, with the goal of modeling the target in a variety of conformations that could be significantly populated under equilibrium conditions.

Modeling is performed using the automodel function of the MODELLER software package~\cite{fiser:prot-sci:2000:modeller,sali:jmb:1993:modeller} to rapidly generate a single model of the target sequence from each template structure.
MODELLER uses simulated annealing cycles along with a minimal forcefield and spatial restraints---generally Gaussian interatomic probability densities extracted from the template structure with database-derived statistics determining the distribution width---to rapidly generate candidate structures of the target sequence from the provided template sequence~\cite{fiser:prot-sci:2000:modeller,sali:jmb:1993:modeller}.

While MODELLER's automodel function can generate its own alignments automatically, a standalone function was preferable for reasons of programming convenience.
As such, we implemented pairwise alignment functionality using the BioPython {\tt pairwise2} module~\cite{cock:2009:biopython}---which uses a dynamic programming algorithm---with the PAM 250 scoring matrix of Gonnet \textit{et al.}~\cite{gonnet:science:1992:exhaustive-matching}, though other choices of scoring matrices available within the module can be selected.
The alignments are carried out with the {\tt align} subcommand, prior to the modeling step which is carried out with the {\tt build\_models} subcommand.
The {\tt align} subcommand also writes a list of the sequence identities for each template to a text file, and this can be used to select models from a desired range of sequence identities.
The {\tt build\_models} subcommand and all subsequent pipeline functions have a {\tt -{}-template\_seqid\_cutoff} flag which can be used to select only models with sequence identities greater than the given value.
We also note that alternative approaches could be used for the alignment stage.
For example, multiple sequence alignment algorithms~\cite{thompson:2011:msa-review}, allow alignments to be guided using sequence data from across the entire protein family of interest, while (multiple) structural alignment algorithms such as MODELLER's {\tt salign} routine~\cite{fiser:prot-sci:2000:modeller,sali:jmb:1993:modeller}, PROMALS3D~\cite{pei:2008:promals3d}, and Expresso and 3DCoffee~\cite{expresso,3dcoffee}, can additionally exploit structural data.
{\bf Ensembler's} modular architecture facilitates the implementation of alternative alignment approaches, and we plan to implement some of these in future versions, to allow exploration of the influence of different alignment methods on model quality.

Models are output as PDB-format coordinate files.
To minimize file storage requirements, {\bf Ensembler} uses the Python {\tt gzip} library to apply compression to all sizeable text files from the modeling stage onwards.
The restraints used by MODELLER could potentially be used in alternative additional refinement schemes, and {\bf Ensembler} thus provides a flag ({\tt -{}-write\_modeller\_restraints\_file}) for optionally saving these restraints to file.
This option is turned off by default, as the restraint files are relatively large (e.g.~$\sim$400 kB per model for protein kinase domain targets), and are not expected to be used by the majority of users.

At this time, the alignment and modeling functions cannot be used to model non-standard amino acids, though we plan to be able to provide this functionality in future versions.
Note that the Ensembler functions for target and template selection only include standard amino acids which match the UniProt canonical isoform sequence, and thus any set of targets and templates selected this way should be compatible with the Ensembler alignment and modeling functions.

\subsection*{Filtering of nearly identical models}

Because {\bf Ensembler} treats individual chains from source PDB structures as individual templates, a number of models may be generated with very similar structures if these individual chains are nearly identical in conformation.
For this reason, and also to allow users to select for high diversity if they so choose, {\bf Ensembler} provides a way to filter out models that are very similar in RMSD.
The {\tt cluster} subcommand can thus be used to identify models which differ from other models in terms of RMSD distance by a user-specified cutoff.
Clustering is performed using the regular spatial clustering algorithm~\cite{noe:jcp:2011:msm-review}, as implemented in the MSMBuilder Python library~\cite{msmbuilder}, which uses mdtraj~\cite{mdtraj} to calculate RMSD (for C$_\alpha$ atoms only) with a fast quaternion characteristic polynomial (QCP)~\cite{theobald:acta-cryst-a:2005:qcp,theobald:j-comput-chem:2010:qcp,theobald:j-comput-chem:2011:qcp} implementation.
A minimum distance cutoff (which defaults to 0.6~\AA) is used to retain only a single model per cluster.

\subsection*{Refinement of models and filtering of poor models by simulation (E)}

A number of refinement methods have been developed to help guide comparative modeling techniques toward more ``native-like'' and physically consistent conformations~\cite{maccallum:prot:2011:casp-refinement,zhang:curr-opin-struct-biol:2009:structure-prediction}.
Both short~\cite{zhang:curr-opin-struct-biol:2009:structure-prediction} and long~\cite{raval:prot:2012:long-timescale-md-refinement} molecular dynamics simulations have been employed for this purpose.
Here, we utilize short molecular dynamic simulations for two purposes: both to slightly relax the initial comparative models and to eliminate those comparative models that result in highly implausible conformations.
This is especially critical here due to the inclusion of even very low sequence identity template structures.
We stress that the limited refinement by molecular simulation here is primarily intended as initial relaxation and filtering stages, where implausible models might cause simulations to immediately fail, crash, or generate implausibly high energies or unstable dynamics.
Exploration of conformational dynamics to derive MSMs, for example, will inevitably require orders of magnitude more simulation effort---very likely tens of microseconds to milliseconds of aggregate dynamics~\cite{pande-beauchamp-bowman:2010:methods:markov-model-review,chodera-noe:curr-opin-struct-biol:2014:msm-review}.

{\bf Ensembler} thus includes a refinement module, which uses short molecular dynamics simulations to refine the models built in the previous step.
As well as improving model quality, this also prepares models for subsequent production MD simulation, including solvation with explicit water molecules, if desired.

Models are first subjected to energy minimization (using the L-BFGS algorithm~\cite{lbfgs}, followed by a short molecular dynamics (MD) simulation with an implicit solvent representation.
This is implemented using the OpenMM molecular simulation toolkit~\cite{eastman:jctc:2012:openmm}, chosen for its flexible Python API, and high performance GPU-acclerated simulation code.
The simulation is run for a default of 100~ps, which in our example applications has been sufficient to filter out poor models (i.e.~those with atomic overlaps unresolved by energy minimization, which result in an unstable simulation), as well as helping to relax model conformations.
As discussed in the Results section, our example application of the {\bf Ensembler} pipeline to the human tyrosine kinase family indicated that of the models which failed implicit solvent MD refinement, the vast majority failed within the first 1~ps of simulation.

The simulation protocol and default parameter values have been chosen to represent current ``best practices'' for the refinement simulations carried out here.
As such, the simulation is performed using Langevin dynamics, with a default force field choice of Amber99SB-ILDN~\cite{amber99sb-ildn}, along with a modified generalized Born solvent model~\cite{gbsa-obc1} as implemented in the OpenMM package~\cite{eastman:jctc:2012:openmm}.
Any of the other force fields or implicit water models implemented in OpenMM can be specified using the {\tt -{}-ff} and {\tt -{}-water\_model} flags respectively.
The simulation length can also be controlled via the {\tt -{}-simlength} flag, and many other important simulation parameters can be controlled from either the API or CLI (via the {\tt -{}-api\_params} flag).
The default values are set as follows---timestep: 2~fs; temperature: 300~K; Langevin collision rate: 20~ps$^{-1}$; pH (used by OpenMM for protonation state assignment): 7.
We also draw attention to a recent paper which indicates that lower Langevin collision rates may result in faster phase space exploration~\cite{basconi:jctc:2013:collision-rates}.

For some studies, it may be useful to specify the protonation states of individual amino acids, rather than rely only on automatic protonation state assignment by OpenMM.
The user can do this by listing the residue numbers and their protonation states in a configuration file ({\tt manual\_overrides.yaml}).
The necessary formatting for the configuration file is specified in the software documentation, and a template file is written when initializing an {\bf Ensembler} project.
Protonation states are specified by naming the appropriate residue variant type in the force field, e.g. {\tt `ASH'} for an aspartic acid residue, as opposed to the aspartate base {\tt `ASP'}.
Any residues which do not have specific protonation states listed in the configuration file will have protonation states assigned automatically by OpenMM.
Note that {\bf Ensembler} currently only supports residue definitions provided by the forcefield definition files---it does not yet have the ability to derive new forcefield parameters for uncommon amino acids, cofactors, or ions not provided by the forcefield.

\subsection*{Solvation and NPT equilibration (F)}

While protein-only models may be sufficient for structural analysis or implicit solvent simulations, {\bf Ensembler} also provides a stage for solvating models with explicit water and performing a round of explicit-solvent MD refinement/equilibration under isothermal-isobaric (NPT) conditions.
The solvation step solvates each model for a given target with the same number of waters to facilitate the integration of data from multiple simulations, which is important for methods such as the construction of MSMs.
The target number of waters is selected by first solvating each model with a specified padding distance (default: 10 \AA), then taking a percentile value from the distribution (default: 68th percentile).
This helps to prevent models with particularly long, extended loops---such as those arising from template structures with unresolved termini---from imposing very large box sizes on the entire set of models.
The TIP3P water model~\cite{tip3p} is used by default, but any of the other explicit water models available in OpenMM, such as TIP4P-Ew~\cite{tip4p-ew}, can be specified using the {\tt -{}-water\_model} flag.
Models are resolvated with the target number of waters by first solvating with zero padding, then incrementally increasing the box size and resolvating until the target is exceeded, then finally deleting sufficient waters to match the target value.
The explicit solvent MD simulation is also implemented using OpenMM, using the Amber99SB-ILDN force field~\cite{amber99sb-ildn} and TIP3P water~\cite{tip3p} by default.
The force field, water model, and simulation length can again be specified using the {\tt -{}-ff}, {\tt -{}-water\_model}, and {\tt -{}-simlength} flags respectively.
Further simulation parameters can be controlled via the API or via the CLI {\tt -{}-api\_params} flag.
Pressure control is performed with a Monte Carlo barostat as implemented in OpenMM, with a default pressure of 1~atm and a period of 50 timesteps.
The remaining simulation parameters have default values set to the same as for the implicit solvent MD refinement.

\subsection*{Model validation with MolProbity score}

{\bf Ensembler} provides a function for validating model quality and filtering models using MolProbity~\cite{molprobity:2007,molprobity:2010}---a widely used tool for validation of protein models, which provides a numerical score derived from features such as steric clashes between atoms, bond geometry, Ramachandran angles, sidechain rotamer outliers, backbone deviations, and the presence of cis-peptides.
This function is accessed via the {\tt validate} subcommand, which for a given target will output a text file containing a list of model IDs sorted by validation score.
The optional {\tt -{}-modeling\_stage} flag specifies which of the three main Ensembler modeling stages to validate---the initial comparative modeling stage, the implicit MD refinement stage, or the explicit MD refinement stage.
If this flag is not used, Ensembler defaults to selecting the latest stage for which models have been generated.
The output text file can be used to filter models based on validation score, for example by using the {\tt package\_models} subcommand.
Protein model validation is a challenging problem and an active area of research for many groups, including the developers of {\bf Ensembler}.
We plan to implement further validation methods in future versions of Ensembler.

\subsection*{Packaging}

{\bf Ensembler} provides a packaging module which can be used to prepare models for subsequent downstream use, such as the use of distributed or cluster computing resources for the generation of MSMs~\cite{pande-beauchamp-bowman:2010:methods:markov-model-review,noe:jcp:2011:msm-review,chodera-noe:curr-opin-struct-biol:2014:msm-review}.
The {\tt package\_models} subcommand currently provides functions (specified via the {\tt -{}-package\_for} flag) for compressing models in preparation for data transfer, or for organizing them with the appropriate directory and file structure for production simulation on the distributed computing platform Folding@home~\cite{shirts-pande:science:2000:fah}.
The module could easily be extended to add methods for preparing models for other purposes.
For example, production simulations could alternatively be run using Copernicus~\cite{pronk:2011:copernicus,pronk:2015:copernicus}---a framework for performing parallel adaptive MD simulations--- or GPUGrid~\cite{buch:2010:gpugrid}---a distributing computing platform which relies on computational power voluntarily donated by the owners of nondedicated GPU-equipped computers. 

An important use of the packaging stage is to filter models based on model quality.
At the current time, the available filtering options are based on either target-template sequence identity or MolProbity validation score.
The {\tt package\_models} subcommand includes optional flags for specifying a sequence identity cutoff (so that only models with a target-template sequence identity above the specified percentage are chosen), a MolProbity validation score cutoff (to choose only models with lower validation scores, which indicate better model quality), or a MolProbity validation score percentile (to choose only models with validation scores lower than the value at the given percentile).

Models can also be exported into trajectory files for the purpose of performing structural analyses across model ensembles using tools like MDTraj~\cite{mdtraj}.
This is done using the {\tt mktraj} subcommand, which writes model coordinates for a given target to a Gromacs~\cite{berendsen:1995:gromacs,lindahl:2001:gromacs} XTC format trajectory (chosen for its wide usage and data compression).
Each frame in the trajectory represents a single model, and  models are sorted in descending order of target-template sequence identity.
Also output for each target are a PDB coordinate file (for use as a topology input file) and a CSV file containing model IDs (in the same order as the frames in the trajectory file) and other data such as target-template sequence identity.
Using the {\tt -{}-modeling\_stage} flag, models can be selected from any of three Ensembler modeling stages - after the initial comparative modeling stage, after implicit MD refinement, or after explicit MD refinement.
If this flag is not used, Ensembler defaults to selecting the latest stage for which models have been generated.

We stress that, despite evidence suggesting that there is a correspondence between solution-state dynamics and structural diversity of related template proteins~\cite{friedland:plos-comput-biol:2009:conformational-dynamics}, all models---especially those derived from low sequence identity templates---are not necessarily representative of conformations thermally accessible to the template proteins of interest.
Care must be exercised in the use and analysis of these models.

\subsection*{Other features}

\paragraph{Tracking provenance information.}
To aid the user in tracking the provenance of each model, each pipeline function also outputs a metadata file, which helps to link data to the software version used to generate it (both {\bf Ensembler} and its dependencies), and also provides timing and performance information, and other data such as hostname.

\paragraph{Rapidly modeling a single template.}
For users interested in simply using {\bf Ensembler} to rapidly generate a set of models for a single template sequence, {\bf Ensembler} provides a command-line tool {\tt quickmodel}, which performs the entire pipeline for a single target with a small number of templates.
For larger numbers of models (such as entire protein families), modeling time is greatly reduced by using the main modeling pipeline, which is parallelized via MPI, distributing computation across each model (or across each template, in the case of the loop reconstruction code), and scaling (in a ``pleasantly parallel'' manner) up to the number of models generated.

\label{section:design}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% RESULTS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Results}
\label{section:results}

\subsection*{Modeling of all human tyrosine kinase catalytic domains}

As a first application of {\bf Ensembler}, we have built models for the human TK family.
TKs (and protein kinases in general) play important roles in many cellular processes and are involved in a number of types of cancer~\cite{krause:nejm:2005:tks-cancer}.
For example, a translocation between the TK Abl1 and the pseudokinase Bcr is closely associated with chronic myelogenous leukemia~\cite{greuber:2013:abl-cancer-review}, while mutations of Src are associated with colon, breast, prostate, lung, and pancreatic cancers~\cite{kim:2009:src-cancer-review}.
Protein kinase domains are thought to have multiple accessible metastable conformation states, and much effort is directed at developing kinase inhibitor drugs which bind to and stabilize inactive conformations~\cite{liu:ncb:2006:kinase-inhibitors-inactive-conformations}.
Kinases are thus a particularly interesting subject for study with MSM methods~\cite{shukla-pande:nature-commun:2014:src-activation-msm}, and this approach stands to benefit greatly from the ability to exploit the full body of available genomic and structural data within the kinase family, e.g.~by generating large numbers of starting configurations to be used in highly parallel MD simulation.

We selected all human TK domains annotated in UniProt as targets, and all available structures of protein kinase domains (of any species) as templates, using the commands shown in Box~\ref{box:tk-gather-targets-templates}.
This returned 93 target sequences and 4433 template structures, giving a total of \num{412269} target-template pairs.
The templates were derived from 3028 individual PDB entries and encompassed 23 different species, with 3634 template structures from human kinase constructs.

The resultant models are available as part of a supplementary dataset which can be downloaded from the Dryad Digital Repository (DOI: \href{http://datadryad.org/review?doi=doi:10.5061/dryad.7fg32}{10.5061/dryad.7fg32}).

Figure~\ref{figure:nstructures-per-tk-target} shows the number of PDB structures available for each of the 93 target TK domains.
While a number of experimental structures are available for some TK domains, many TKs have few or no structures. 
{\bf Ensembler} thus helps to overcome this unequal distribution of structural information when building protein models for simulation by exploiting homologous structural data from a wider range of protein kinase domains and species.

\begin{figure}[h]
    % \includegraphics[width=\textwidth]{nstructures_per_tk_target/nstructures_per_tk_target.pdf}
    \caption{{\bf Number of PDB structures available for each TK target.}
    Data is shown for each of the 93 TK kinase domains, sorted in order of the number of available PDB structures for each domain. 
    The labels indicate the UniProt name for the target protein plus an index for the kinase domain (three of the selected proteins have two kinase domains). 
    Each PDB chain is counted individually, and only chains which contain the target domain are counted.
}
    \label{figure:nstructures-per-tk-target}
\end{figure}

\subsection*{Ensembler modeling statistics}

Crystallographic structures of kinase catalytic domains generally contain a significant number of missing residues (median 11, mean 14, standard deviation 13, max 102) due to the high mobility of several loops (Fig.~\ref{figure:loopmodel-nmissing-residues}, top), with a number of these missing spans being significant in length (median 5, mean 7, standard deviation 6, max 82; Fig.~\ref{figure:loopmodel-nmissing-residues}, bottom).
To reduce the reliance on the MODELLER rapid model construction stage to reconstruct very long unresolved loops, unresolved template residues were first remodeled using the {\tt loopmodel} subcommand.
Out of \num{3666} templates with one or more missing residues, \num{3134} were successfully remodeled by the Rosetta loop modeling stage (with success defined simply as program termination without error); most remodeling failures were attributable to unsatisfiable spatial constraints imposed by the original template structure.
There was some correlation between remodeling failures and the number of missing residues (Fig.~\ref{figure:loopmodel-nmissing-residues}, top); templates for which remodeling failed had a median of 20 missing residues, compared to a median of 14 missing residues for templates for which remodeling was successful (when excluding templates with no missing residues).

\begin{figure}[h]
    % \includegraphics[width=\textwidth]{loopmodel_analysis/nmissing_resis_distributions.pdf}
    \caption{{\bf Distributions for the number of missing residues in the TK templates.}
    \emph{Upper:} The number of missing residues per template, for all templates (blue) and for only those templates for which template remodeling with the {\tt loopmodel} subcommand failed (red).
    Templates for which remodeling failed had a median of 20 missing residues, compared to a median of 14 missing residues for templates for which remodeling was successful.
    \emph{Lower:} The number of residues in each missing loop, for all templates.    
}
    \label{figure:loopmodel-nmissing-residues}
\end{figure}

Following loop remodeling, the {\bf Ensembler} pipeline was performed up to and including the implicit solvent MD refinement stage, which completed with \num{389067} (94\%) surviving models across all TKs.
To obtain statistics for the solvation stage without generating a sizeable amount of coordinate data (with solvated PDB coordinate files taking up about 0.9 MB each), the {\tt solvate} subcommand was performed for two representative individual kinases (\emph{Src} and \emph{Abl1}).

The number of models which survived each stage are shown in Fig.~\ref{figure:pipeline}, indicating that the greatest attrition occurred during the modeling stage.
The number of refined models for each target ranged from \num{4046} to \num{4289}, with a median of \num{4185}, mean of \num{4184}, and standard deviation of \num{57}.
Fig.~\ref{figure:pipeline} also indicates the typical timing achieved on a cluster for each stage, showing that the {\tt build\_models} and {\tt refine\_implicit\_md} stages are by far the most compute-intensive.

The files generated for each model (up to and including the implicit solvent MD refinement stage) totaled $\sim$116~kB in size, totalling 0.5~GB per TK target or 42~GB for all~93 targets.
The data generated per model breaks down as 39~kB for the output from the modeling stage (without saving MODELLER restraints files, which are about 397~kB per model) and 77~kB for the implicit solvent MD refinement stage.

\subsection*{Evaluation of model quality and utility}

\paragraph{All tyrosine kinases.}

To evaluate the variety of template sequence similarities relative to each target sequence, we calculated sequence identity distributions, as shown in Fig.~\ref{figure:sequence-identity-distribution}.
This suggests an intuitive division into three categories, with \num{355712} models in the 0--35\% sequence identity range, \num{51330} models in the 35--55\% range, and \num{5227} models in the 55--100\% range.
We then computed the RMSD distributions for the models created for each target (relative to the model derived from the template with highest sequence identity) Fig.~\ref{figure:rmsd-distributions}, to assess the diversity of conformations captured by the modeling pipeline.
Furthermore, to understand the influence of sequence identity on the conformational similarities of the resulting models, the RMSD distributions were stratified based on the three sequence identity categories described above.
This analysis indicates that higher sequence identity templates result in models with lower RMSDs, while templates with remote sequence identities result in larger RMSDs on average, recapitulating the observation made years ago by Chothia and Lesk~\cite{chothia-lesk:embo-j:1986:rmsd-sequence-identity-relation}.

We also used the {\tt ensembler validate} subcommand to subject the refined models to analysis with MolProbity~\cite{molprobity:2007,molprobity:2010}.
\begin{figure}[h]
    % \includegraphics[width=1.0\columnwidth]{seqid_dist/seqid_dist.pdf}

    \caption{{\bf Template-target sequence identity distribution for human tyrosine kinase catalytic domains.}
    Sequence identities are calculated from all pairwise target-template alignments, where targets are human kinase catalytic domain sequences and templates are all kinase catalytic domains from any organism with structures in the PDB, as described in the text.
    A kernel density estimate of the target-template sequence identity probability density function is shown as a solid line with shaded region, while the corresponding cumulative distribution function is shown as a dashed line.
    }
  \label{figure:sequence-identity-distribution}
\end{figure}

\begin{figure}[h]
    % \centering
    % {\label{figure:rmsd-distribution-joint}%
    %    \includegraphics[height=0.6\textwidth]{rmsddist/rmsddist2-joint.pdf}
    % }

    % {\label{figure:rmsd-distribution-by-sequence-identity}%
    %    \includegraphics[height=0.6\textwidth]{rmsddist/rmsddist2.pdf}
    % }

    \caption{{\bf Distribution of RMSDs to all TK catalytic domain models relative to the model derived from the highest sequence identity template.}    
    Distributions are built from data from all 93 TK domain targets.
    To better illustrate how conformational similarity depends on sequence identity, the lower plot illustrates the distributions as stratified into three sequence identity classes: high identity (55--100\%), moderate identity (35--55\%), and remote identity (0--35\%).
    The plotted distributions have been smoothed using kernel density estimation.
  }
  \label{figure:rmsd-distributions}
\end{figure}

The MolProbity scores varied from 0.92 to 4.80, with a median of 3.84, a mean of 3.22, and a standard deviation of 1.07.
Lower numbers represent better quality models.
When stratified by the same sequence identity ranges as above, the mean scores were as follows: 2.96 (55--100\% sequence identity), 3.13 (35--55\% sequence identity), 3.24 (0--35\% sequence identity).
This indicates that models with lower target-template sequence identities tend to be lower quality according to MolProbity analysis, as would be expected.

We also analyzed the potential energies of the models at the end of the implicit solvent MD refinement stage.
These ranged from -14180~kT to -3160~kT, with a median of -9501~kT, mean of -9418~kT, and a standard deviation of 1198~kT (with a simulation temperature of 300~K).
The distributions---stratified using the same sequence identity ranges again---are plotted in Fig.~\ref{figure:energies-implicit}, indicating that higher sequence identity templates tend to result in slightly lower energy models.
Of the \num{4973} models which failed to complete the implicit refinement MD stage, all except 9 failed within the first 1~ps of simulation.

\begin{figure}[h]
    % \includegraphics[width=1.0\columnwidth]{energies/energies.pdf}

    \caption{{\bf Distribution of final energies from implicit solvent MD refinement of TK catalytic domain models.}
    To illustrate how the energies are affected by sequence identity, the models are separated into three sequence identity classes: high identity (55--100\%), moderate identity (35--55\%), and remote identity (0--35\%).
    The plotted distributions have been smoothed using kernel density estimation.
    Refinement simulations were carried out at the default temperature of 300~K.
  }
  \label{figure:energies-implicit}
\end{figure}

\paragraph{Src and Abl1.}

To provide a more detailed evaluation of the variety and utility of generated models, we have analyzed two specific TKs (\emph{Src} and \emph{Abl1}) in depth.
Due to their importance in cancer, these kinases have been the subject of numerous detailed structural and simulation studies.
In terms of structural data, a large number of crystal structures have been solved (with or without ligands such as nucleotide substrate mimetics or small-molecule inhibitors), revealing a variety of conformations accessible to these kinases.
A recent large-scale MSM study has also studied the activation pathway of Src~\cite{shukla-pande:nature-commun:2014:src-activation-msm}, while a separate study employed biased sampling techniques to dissect the role of conformational changes in selectivity and affinity of imatinib recognition of Abl~\cite{roux:pnas:2013:abl-imatinib}.

\paragraph{Visualizing model structural diversity.}
Fig.~\ref{figure:superposition} shows a superposition of a set of representative models of \emph{Src} and \emph{Abl1}.
Models were first stratified into three ranges, based on the structure of the sequence identity distribution (Fig.~\ref{figure:sequence-identity-distribution}), then subjected to RMSD-based $k$-medoids clustering (using the msmbuilder clustering package~\cite{msmbuilder}) to pick three representative models from each sequence identity range.
Each model is colored and given a transparency based on the sequence identity between the target and template sequence.
The figure gives an idea of the variance present in the generated models.
High sequence identity models (in opaque blue) tend to be quite structurally similar, with some variation in loops or changes in domain orientation.

\begin{figure}[h]
    % \includegraphics[width=1.0\columnwidth]{superposition-src_abl/superposed-seqid_classes-clustered-one_fig}
    
    \caption{{\bf Superposition of clustered models of Src and Abl1.}
    Superposed renderings of nine models each for Src and Abl1, 
    giving some indication the diversity of conformations generated by Ensembler.
    The models for each target were divided into three sequence identity ranges (as in Fig.~\ref{figure:rmsd-distributions}), and RMSD-based $k$-medoids clustering was performed (using the msmbuilder clustering package~\cite{msmbuilder}) to select three clusters from each.
    The models shown are the centroids of each cluster.
    Models are colored and given transparency based on their sequence identity, so that high sequence identity models are blue and opaque, while lower sequence identity models are transparent and red.
  }
  \label{figure:superposition}
\end{figure}

The Abl1 renderings in Fig.~\ref{figure:superposition} indicate one high sequence identity model with a long unstructured region at one of the termini, which was unresolved in the original template structure.
While such models are not necessarily incorrect or undesirable, it is important to be aware of the effects they may have on production simulations performed under periodic boundary conditions, as long unstructured termini can be prone to interact with a protein's periodic image.
Lower sequence identity models (in transparent white or red) indicate much greater variation in all parts of the structure.
We believe the mix of high and low sequence identity models to be particularly useful for methods such as MSM building, which require thorough sampling of the conformational landscape.
The high sequence identity models could be considered to be the most likely to accurately represent true metastable states.
Conversely, the lower sequence identity models could be expected to help push a simulation into regions of conformation space which might take intractably long to reach if starting a single metastable conformation.

\paragraph{Comparison with known biochemically relevant conformations.}
To evaluate the models of \emph{Src} and \emph{Abl1} in the context of the published structural biology literature on functionally relevant conformations, we have focused on two residue pair distances thought to be important order parameters for the regulation of protein kinase domain activity.
We use the residue numbering schemes for chicken Src (commonly employed in the literature even in reference to human Src)~\cite{xu:1999:2src, cowan-jacob:2005:1y57} and human Abl1 isoform A~\cite{young:2006:2f4j, cowan-jacob:2006:2hyy, levinson:2006:2g1t} respectively; see \nameref{S1_Text} for the exact sequences and numbering schemes used.

Fig.~\ref{figure:src-ref-structures} shows two structures of \emph{Src} believed to represent inactive (PDB code: 2SRC)~\cite{xu:1999:2src} and active (PDB code: 1Y57)~\cite{cowan-jacob:2005:1y57} states.
One notable feature which distinguishes the two structures is the transfer of an electrostatic interaction of E310 from R409 (in the inactive state) to K295 (in the active state), brought about by a rotation of the $\alpha$C-helix.
These three residues are also well conserved~\cite{kannan:jmb:2005:kinase-regulation}, and a number of experimental and simulation studies have suggested that this electrostatic switching process plays a role in a regulatory mechanism shared across the protein kinase family~\cite{shukla-pande:nature-commun:2014:src-activation-msm, foda:nat-comm:2015:src, ozkirimli:prot-sci:2008:src}.
As such, we have plotted the distance between these two residue pairs for the {\bf Ensembler} models for \emph{Src} and \emph{Abl1}, as well as \emph{Flt4} (Fig.~\ref{figure:pair-distances}).
The models all show strong coverage of regions in which either of the electrostatic interactions is fully formed (for models across all levels of target-template sequence identity), as well as a wide range of regions in-between (mainly models with low sequence identity).
We thus expect that such a set of models, if used as starting configurations for highly parallel MD simulation, could greatly aid in sampling of functionally relevant conformational states.
The \emph{Flt4} models (Fig.~\ref{figure:src-ref-structures}c) are of particular note, as there are no available crystal structures of the kinase domain of this TK protein (which is involved in tumor angiogenesis and lymphangiogenesis~\cite{Su:2006:flt4cancer}), yet the models generated here include structural motifs which are conserved and of known importance to other proteins of the same family.

\begin{figure}[h]
    % \includegraphics[width=1.0\textwidth]{residue_pair_distances/src/src_ref_structures}

    \caption{{\bf Two structures of Src, indicating certain residues involved in activation.}
    In the inactive state, E310 forms a salt bridge with R409.
    During activation, the $\alpha$C-helix (green) moves and rotates, orienting E310 towards the ATP-binding site and allowing it to instead form a salt bridge with K295.
    This positions K295 in the appropriate position for catalysis.
    Note that ANP (phosphoaminophosphonic acid-adenylate ester; an analog of ATP) is only physically present in the 2SRC structure.
    To aid visualization of the active site in 1Y57, it has been included in the rendering by structurally aligning the surrounding homologous protein residues.
  }
  \label{figure:src-ref-structures}
\end{figure}

\begin{figure}[h]

    % \subfloat[Src]{\label{figure:src-distances}%
    %    \includegraphics[height=0.31\textwidth]{residue_pair_distances/src_abl_flt4_combined/distances_src.png}
    % }
    % \subfloat[Abl]{\label{figure:abl-distances}%
    %    \includegraphics[height=0.31\textwidth]{residue_pair_distances/src_abl_flt4_combined/distances_abl.png}
    % }
    % \subfloat[Flt4]{\label{figure:flt4-distances}%
    %    \includegraphics[height=0.31\textwidth]{residue_pair_distances/src_abl_flt4_combined/distances_flt4.png}
    % }

    \caption{{\bf Src, Abl1, and Flt4 models projected onto the distances between two conserved residue pairs, colored by sequence identity.}
    Two Src structures (PDB entries 1Y57~\cite{cowan-jacob:2005:1y57} and 2SRC~\cite{xu:1999:2src}) are projected onto the plots for reference, representing active and inactive states respectively.
    These structures and the residue pairs analyzed here are depicted in Fig.~\ref{figure:src-ref-structures}.
    Distances are measured between the center of masses of the three terminal sidechain heavy atoms of each residue.
    The atom names for these atoms, according to the PDB coordinate files for both reference structures, are---Lys: NZ, CD, CE (ethylamine); Glu: OE1, CD, OE2 (carboxylate); Arg: NH1, CZ, NH2 (part of guanidine).
    }
  \label{figure:pair-distances}
\end{figure}

\section*{Availability and Future Directions}
\label{section:availability}

\subsection*{Availability}

The code for {\bf Ensembler} is hosted on the collaborative open source software development platform GitHub (\href{http://github.com/choderalab/ensembler}{github.com/choderalab/ensembler}).
The latest release can be installed via the {\tt conda} package manager for Python (\href{http://conda.pydata.org}{conda.pydata.org}), using the commands shown in Box~\ref{box:installation}.
This will install all required dependencies, though a license must first be obtained for MODELLER.
The optional dependencies Rosetta and MolProbity are not available through the {\tt conda} package manager, and thus must be installed separately by the user according to the instructions for those packages..
The latest source can be downloaded from the GitHub repository, which also contains up-to-date instructions for building and installing the code.
Documentation can be found at \href{http://ensembler.readthedocs.org/en/latest/}{ensembler.readthedocs.org}.

\begin{codebox}[h]
\begin{adjustwidth}{-2.25in}{0in}
\footnotesize
\begin{Verbatim}[frame=single]
export KEY_MODELLER="Modeller license key"
conda config -add channels https://conda.anaconda.org/omnia
conda config -add channels https://conda.anaconda.org/salilab
conda install ensembler
\end{Verbatim}
\caption{{\bf Ensembler installation using {\tt conda} in a {\tt bash} shell. A license for MODELLER must first be obtained, and the license key should be used in place of the text {\tt "Modeller license key"}.}
}
\label{box:installation}
\end{adjustwidth}
\end{codebox}

A supplementary dataset can also be downloaded from the Dryad Digital Repository (DOI: \href{https://dx.doi.org/10.5061/dryad.7fg32}{10.5061/dryad.7fg32}).
This contains the TK models described in the \ref{section:results} section, general information on the targets and templates, plus a script and instructions for regenerating the same dataset.

\subsection*{Future Directions}

We recognize that the current version of {\bf Ensembler} has a number of limitations that bound its domain of applicability: support for nonnatural amino acids is currently rudimentary and confined to those already appearing in the forcefield; cofactors cannot currently be automatically modeled in; ligands, cofactors, and nonnatural amino acids cannot yet be automatically parameterized; protonation state assignment is limited to selection of the most populated state based on the intrinsic p$K_a$ or user-specified overrides; the modeling of missing loops is rudimentary, relying on the subsequent dynamics for relaxation; there is not yet support for modeling of distinct domains from different templates, or the use of multiple templates to model a single domain.
Nevertheless, there are a great number of use cases for this first version of an automated tool for simulation preparation at the superfamily scale.
To expand this domain of applicability, there are a number of obvious additions and improvements which we plan to implement in future versions of {\bf Ensembler}.

\paragraph{Template remodeling.}
The lack of crystallographically-resolved regions of template structures presents a challenge to deriving structures from these templates by comparative modeling, especially in kinases, where loops are frequently unresolved.
Improvements over the Rosetta-based strategy described here are likely possible, especially given the number of modeling failures observed in the template refinement stage (Fig.~\ref{figure:loopmodel-nmissing-residues}).
An alternative approach could be to re-refine complete-chain template structures to the experimentally-derived electron density or scattering data deposited in the RCSB using methods capable of exploiting the scattering data and crystallographic symmetry~\cite{schnieders:biophys-j:2010:x-ray-refinement}.
Even if definitive placement of these unresolved regions is impossible, plausible locations constrained by weak scattering data and strong steric exclusion of crystallographic neighbors may provide a great deal of useful information, especially when combined with forcefield priors~\cite{schnieders:jctc:2011:space-group-pme}.

\paragraph{Comparative modeling.}
Comparative protein modeling can be approached in a number of different ways, with varying degrees of complexity.
The comparative modeling stage of {\bf Ensembler} currently uses MODELLER, but a number of excellent alternatives---such as RosettaCM~\cite{baker:structure:2013:rosettacm} and the I-TASSER Suite~\cite{yang:nature-methods:2015:i-tasser-suite}---can be added as user-selectable alternative choices.
Additional options could be added to allow more expensive loop-modeling approaches to be employed to handle long insertions.

\paragraph{Protonation states.}
Some amino acids can exist in different protonation states, depending on pH and on their local environment.
These protonation states can have important effects on biological processes.
For example, long timescale MD simulations have suggested that the conformation of the DFG motif of the TK Abl1---believed to be an important regulatory mechanism~\cite{nagar:cell:2003:abl-1opl}---is controlled by protonation of the aspartate~\cite{shan:pnas:2009:abl}.
Currently, protonation states are assigned simply based on pH (a user-controllable parameter).
At neutral pH, histidines have two protonation states which are approximately equally likely, and in this situation the selection is therefore made based on which state results in a better hydrogen bond.
It would be highly desirable to instead use a method which assigns amino acid protonation states based on a rigorous assessment of the local environment.
We thus plan to implement an interface and command-line function for assigning protonation states with MCCE2~\cite{alexov-gunner:biophys-j:1997:mcce2,georgescu-alexov-gunner:biophys-j:2002:mcce2,song-mao-gunner:j-comput-chem:2009:mcce2}, which uses electrostatics calculations combined with Monte Carlo sampling of side chain conformers to calculate pKa values.

\paragraph{Cofactors, structural ions, and ligands.}
Many proteins require the presence of various types of non-protein atoms and molecules for proper function, such as metal ions (e.g.~Mg$^{+2}$), cofactors (e.g.~ATP) or post-translational modifications (e.g.~phosphorylation, methylation, glycosylation, etc.), and we thus plan for {\bf Ensembler} to eventually have the capability to include such entities in the generated models.
Binding sites for metal ions are frequently found in proteins, often playing a role in catalysis.
For example, protein kinase domains contain two binding sites for divalent metal cations, and display significantly increased activity in the presence of Mg$^{2+}$~\cite{adams:prot-sci:1993:kinase-metal-ions}, the divalent cation with highest concentration in mammalian cells.
Metal ions are often not resolved in experimental structures of proteins, but by taking into account the full range of available structural data, it should be possible in many cases to include metal ions based on the structures of homologous proteins.
We are careful to point out, however, that metal ion parameters in classical MD force fields have significant limitations, particularly in their interactions with proteins~\cite{sousa:book:2010:metal-ion-parameters-difficulties}.
Cofactors and post-translational modifications are also often not fully resolved in experimental structures, and endogenous cofactors are frequently substituted with other molecules to facilitate experimental structural analysis.
Future extensions to {\bf Ensembler} could transfer cofactor and ion coordinates from homologous proteins in which these components are resolved.

\paragraph{Post-translationally modified amino acids and other molecules without forcefield parameters.}
A major challenge in the preparation of simulations of proteins of interest is the wide variety of post-translational modifications possible that are often functionally or structurally relevant.
Often, forcefields lack parameters for these residues, or for other cofactors or ligands that might be vital to probing the relevant structural dynamics of these systems.
While tools such as Antechamber~\cite{gaff,gaff2} can rapidly generate small molecule parameters in an automated manner, the parameterization of polymeric residues or covalently attached cofactors is much more challenging.
In addition, small molecule forcefields are generally tied to specific corresponding protein and nucleic acid forcefields, meaning that different procedures may be needed to generate consistent parameters.

\paragraph{Long insertions and deletions.}
Another limitation with the present version of {\bf Ensembler} involves the treatment of members of a protein family with especially long residue insertions or deletions.
For example, the set of all human protein kinase domains listed in UniProt have a median length of 265 residues (mean 277) and a standard deviation of 45, yet the minimum and maximum lengths are 102 and 801 respectively.
The latter value corresponds to the protein kinase domain of serine/threonine-kinase \emph{greatwall}, which includes a long insertion between the two main lobes of the catalytic domain.
In principle, such insertions could be excluded from the generated models, though a number of questions would arise as to how best to approach this.

\paragraph{Markov state model (MSM) construction and model utility.}
We are actively utilizing {\bf Ensembler}-generated models to seed the construction of Markov state models (MSMs)~\cite{pande-beauchamp-bowman:2010:methods:markov-model-review,chodera-noe:curr-opin-struct-biol:2014:msm-review}.
While the observation that high sequence identity templates are likely to reflect accessible solution-phase conformations suggests that a number of these models occupy thermally accessible regions of configuration space~\cite{friedland:plos-comput-biol:2009:conformational-dynamics}, many models---especially those derived from very low sequence identity templates---are likely to be highly unrepresentative of conformations populated at equilibrium by the target protein.
It is likely that even with hundreds of microseconds to milliseconds of aggregated dynamics, many of these poor quality models will remain trapped in inaccessible and irrelevant regions of configuration space.
Standard approaches to MSM construction now employ an \emph{ergodic trimming} step~\cite{caflisch:jpcb:2011:msm-subgraph,msmbuilder} to prune away disconnected minor regions of configuration space, and this step is expected to be essential in the successful construction of MSMs using {\bf Ensembler}-derived models.

\subsection*{Conclusion}

We believe {\bf Ensembler} to be an important first step toward enabling computational modeling and simulation of proteins on the scale of entire protein families, and suggest that it could likely prove useful for tasks beyond its original aim of providing diverse starting configurations for MD simulations.
The code is open source and has been developed with extensibility in mind, in order to facilitate its customization for a wide range of potential uses by the wider scientific community.











Sed non aliquet felis. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris commodo justo ac dui pretium imperdiet. Sed suscipit iaculis mi at feugiat. Ut neque ipsum, luctus id lacus ut, laoreet scelerisque urna. Phasellus venenatis, tortor nec vestibulum mattis, massa tortor interdum felis, nec pellentesque metus tortor nec nisl. Ut ornare mauris tellus, vel dapibus arcu suscipit sed. Nam condimentum sem eget mollis euismod. Nullam dui urna, gravida venenatis dui et, tincidunt sodales ex. Nunc est dui, sodales sed mauris nec, auctor sagittis leo. Aliquam tincidunt, ex in facilisis elementum, libero lectus luctus est, non vulputate nisl augue at dolor. For more information, see \nameref{S1_Text}.

\section*{Supporting Information}

\subsection*{S1 Text}
\label{S1_Text}
{\bf Sequences and residue numbering schemes for Src and Abl1.}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% ACKNOWLEDGMENTS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgments}
\label{section:acknowledgments}

The authors are grateful to Robert McGibbon (Stanford) and Arien S. Rustenburg (MSKCC) for many excellent software engineering suggestions.
The authors thank Nicholas M.~Levinson (University of Minnesota), Markus A.~Seeliger (Stony Brook), Diwakar Shukla (Stanford), and Avner Schlessinger (Mount Sinai) for helpful scientific feedback on modeling kinases.
The authors are grateful to Benjamin Webb and Andrej \v{S}ali (UCSF) for help with the MODELLER package, Peter Eastman and Vijay Pande (Stanford) for assistance with OpenMM, and Marilyn Gunner (CCNY) for assistance with MCCE2, as well as the anonymous referees for constructive feedback on this manuscript.
All authors acknowledge support from the Sloan Kettering Institute.
JDC, KAB, and DLP acknowledge partial support from NIH grant P30 CA008748.
JDC and DLP also acknowledge the generous support of a Louis V.~Gerstner Young Investigator Award.
KAB was also supported in part by Starr Foundation grant I8-A8-058.
PBG acknowledges partial funding support from the Weill Cornell Graduate School of Medical Sciences.

\nolinenumbers

%\section*{References}
% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% OR
%
% Compile your BiBTeX database using our plos2015.bst
% style file and paste the contents of your .bbl file
% here.
% 

% DONE - when ready, comment line below and uncomment subsequent lines. Then paste in formatted bibliography from ensembler.bbl
% \bibliography{ensembler.bib}

\begin{thebibliography}{100}

\bibitem{craik:science:2009:trapping-moving-targets}
Lee GM, Craik CS.
\newblock Trapping moving targets with small molecules.
\newblock Science. 2009;324:213.

\bibitem{eastman:jctc:2012:openmm}
Eastman P, Friedrichs MS, Chodera JD, Radmer RJ, Bruns CM, Ku JP, et~al.
\newblock {OpenMM} 4: A Reusable, Extensible, Hardware Independent Library for
  High Performance Molecular Simulation.
\newblock J Chem Theory Comput. 2012;9(1):461--469.

\bibitem{walker:jctc:2013:amber-gpu}
Salomon-Ferrer R, G\"{o}tz AW, Poole D, Grand SL, Walker RC.
\newblock Routine microsecond molecular dynamics simulations with {AMBER} on
  {GPUs}. 2. Explicit solvent particle mesh {Ewald}.
\newblock J Chem Theor Comput. 2013;9(9):3878--3888.

\bibitem{shirts-pande:science:2000:fah}
Shirts M, Pande VS.
\newblock Screen {Savers} of the {World} {Unite}!
\newblock Science. 2000 Dec;290(5498):1903--1904.

\bibitem{pronk:2011:copernicus}
Pronk S, Larsson P, Pouya I, Bowman GR, Haque IS, Beauchamp K, et~al.
\newblock Copernicus: {A} {New} {Paradigm} for {Parallel} {Adaptive}
  {Molecular} {Dynamics}.
\newblock In: Proceedings of 2011 {International} {Conference} for {High}
  {Performance} {Computing}, {Networking}, {Storage} and {Analysis}. {SC} '11.
  New York, NY, USA: ACM; 2011. p. 60:1--60:10.

\bibitem{pronk:2015:copernicus}
Pronk S, Pouya I, Lundborg M, Rotskoff G, Wes\'{e}n B, Kasson PM, et~al.
\newblock Molecular {Simulation} {Workflows} as {Parallel} {Algorithms}: {The}
  {Execution} {Engine} of {Copernicus}, a {Distributed} {High}-{Performance}
  {Computing} {Platform}.
\newblock Journal of Chemical Theory and Computation. 2015 Apr;.

\bibitem{buch:2010:gpugrid}
Buch I, Harvey MJ, Giorgino T, Anderson DP, De~Fabritiis G.
\newblock High-{Throughput} {All}-{Atom} {Molecular} {Dynamics} {Simulations}
  {Using} {Distributed} {Computing}.
\newblock Journal of Chemical Information and Modeling. 2010
  Mar;50(3):397--403.

\bibitem{pande-beauchamp-bowman:2010:methods:markov-model-review}
Pande VS, Beauchamp K, Bowman GR.
\newblock Everything you wanted to know about {Markov State Models} but were
  afraid to ask.
\newblock Methods. 2010;52:99--105.

\bibitem{noe:jcp:2011:msm-review}
Prinz JH, Wu H, Sarich M, Keller B, Fischbach M, Held M, et~al.
\newblock Markov models of molecular kinetics: Generation and validation.
\newblock J Chem Phys. 2011;134:174105.

\bibitem{chodera-noe:curr-opin-struct-biol:2014:msm-review}
Chodera JD, No\'{e} F.
\newblock Markov state models of biomolecular conformational dynamics.
\newblock Curr Opin Struct Biol. 2014;25:135--144.

\bibitem{moult:proteins:2014:casp10}
Moult J, Fidelis K, Kryshtafovych A, Schwede T, Tramontano A.
\newblock Critical assessment of methods of protein structure prediction
  ({CASP}) — round x.
\newblock Proteins: Structure, Function, and Bioinformatics. 2014 Feb;82:1--6.

\bibitem{baker:science:2001:structure-prediction-review}
Baker D, {\v{S}ali} A.
\newblock Protein {Structure} {Prediction} and {Structural} {Genomics}.
\newblock Science. 2001 Oct;294(5540):93--96.

\bibitem{baker:structure:2013:rosettacm}
Song Y, {DiMaio} F, Wang RYR, Kim D, Miles C, Brunette T, et~al.
\newblock High-resolution comparative modeling with {RosettaCM}.
\newblock Structure. 2013;21:1735--1742.

\bibitem{yang:nature-methods:2015:i-tasser-suite}
Yang J, Yan R, Roy A, Xu D, Poisson J, Zhang Y.
\newblock The {I-TASSER Suite}: Protein structure and function prediction.
\newblock Nature Methods. 2015;12:7.

\bibitem{daggett:structure:2010:dynameomics}
{van der Kamp} MW, Schaeffer RD, Jonsson AL, Scouras AD, Simms AM, Toofanny RD,
  et~al.
\newblock Dynameomics: A comprehensive database of protein dynamics.
\newblock Structure. 2010;18:423--435.

\bibitem{friedland:plos-comput-biol:2009:conformational-dynamics}
Friedland GD, Lakomek NA, Griesinger C, Meiler J, Kortemme T.
\newblock A correspondence between solution-state dynamics of an individual
  protein and the sequence and conformational diversity of its family.
\newblock PLoS Comput Biol. 2009;5:e1000393.

\bibitem{sali:2012:allostery-modeling}
Weinkam P, Pons J, Sali A.
\newblock Structure-based model of allostery predicts coupling between distant
  sites.
\newblock Proceedings of the National Academy of Sciences of the United States
  of America. 2012 Mar;109(13):4875--4880.

\bibitem{msmbuilder}
Beauchamp KA, Bowman GR, Lane TJ, Maibaum L, Haque IS, Pande VS.
\newblock {MSMBuilder}2: {Modeling} {Conformational} {Dynamics} on the
  {Picosecond} to {Millisecond} {Scale}.
\newblock Journal of Chemical Theory and Computation. 2011
  Oct;7(10):3412--3419.

\bibitem{caflisch:jpcb:2011:msm-subgraph}
Scalco R, Caflisch A.
\newblock Equilibrium distribution from distributed computing (simulations of
  protein folding).
\newblock The Journal of Physical Chemistry B. 2011 May;115(19):6358--6365.

\bibitem{uniprot:2015}
Consortium TU.
\newblock {UniProt}: a hub for protein information.
\newblock Nucleic Acids Research. 2015 Jan;43(D1):D204--D212.

\bibitem{velankar:2013:sifts}
Velankar S, Dana JM, Jacobsen J, van Ginkel G, Gane PJ, Luo J, et~al.
\newblock {SIFTS}: {Structure} {Integration} with {Function}, {Taxonomy} and
  {Sequences} resource.
\newblock Nucleic Acids Research. 2013 Jan;41(D1):D483--D489.

\bibitem{qian:nature:2007:modeller}
Qian B, Raman S, Das R, Bradley P, McCoy AJ, Read RJ, et~al.
\newblock High-resolution structure prediction and the crystallographic phase
  problem.
\newblock Nature. 2007 Nov;450(7167):259--264.

\bibitem{wang:jmb:2007:modeller}
Wang C, Bradley P, Baker D.
\newblock Protein–{Protein} {Docking} with {Backbone} {Flexibility}.
\newblock Journal of Molecular Biology. 2007 Oct;373(2):503--519.

\bibitem{fiser:prot-sci:2000:modeller}
Fiser A, Do RKG, {\v{S}ali} A.
\newblock Modeling of loops in protein structures.
\newblock Protein Science. 2000 Jan;9(9):1753--1773.

\bibitem{sali:jmb:1993:modeller}
{\v{S}ali} A, Blundell TL.
\newblock Comparative {Protein} {Modelling} by {Satisfaction} of {Spatial}
  {Restraints}.
\newblock Journal of Molecular Biology. 1993 Dec;234(3):779--815.

\bibitem{cock:2009:biopython}
Cock PJA, Antao T, Chang JT, Chapman BA, Cox CJ, Dalke A, et~al.
\newblock Biopython: freely available {Python} tools for computational
  molecular biology and bioinformatics.
\newblock Bioinformatics (Oxford, England). 2009 Jun;25(11):1422--1423.

\bibitem{gonnet:science:1992:exhaustive-matching}
Gonnet GH, Cohen MA, Brenner SA.
\newblock Exhaustive matching of the entire protein sequence database.
\newblock Science. 1992;256:1443--1445.

\bibitem{thompson:2011:msa-review}
Thompson JD, Linard B, Lecompte O, Poch O.
\newblock A {Comprehensive} {Benchmark} {Study} of {Multiple} {Sequence}
  {Alignment} {Methods}: {Current} {Challenges} and {Future} {Perspectives}.
\newblock PLoS ONE. 2011 Mar;6(3):e18093.

\bibitem{pei:2008:promals3d}
Pei J, Kim BH, Grishin NV.
\newblock {PROMALS}3D: a tool for multiple protein sequence and structure
  alignments.
\newblock Nucleic Acids Research. 2008 Apr;36(7):2295--2300.

\bibitem{expresso}
Armougom F, Moretti S, Poirot O, Audic S, Dumas P, Schaeli B, et~al.
\newblock Expresso: automatic incorporation of structural information in
  multiple sequence alignments using 3D-{Coffee}.
\newblock Nucleic Acids Research. 2006 Jul;34(Web Server issue):W604--608.

\bibitem{3dcoffee}
Poirot O, Suhre K, Abergel C, O'Toole E, Notredame C.
\newblock 3DCoffee@igs: a web server for combining sequences and structures
  into a multiple sequence alignment.
\newblock Nucleic Acids Research. 2004 Jul;32(Web Server issue):W37--40.

\bibitem{mdtraj}
McGibbon RT, Beauchamp KA, Schwantes CR, Wang LP, Hern{\'a}ndez CX, Harrigan
  MP, et~al.
\newblock MDTraj: a modern, open library for the analysis of molecular dynamics
  trajectories.
\newblock bioRxiv. 2014;.

\bibitem{theobald:acta-cryst-a:2005:qcp}
Theobald DL.
\newblock Rapid calculation of {RMSDs} using a quaternion-based characteristic
  polynomial.
\newblock Acta Cryst A. 2005;61:478--480.

\bibitem{theobald:j-comput-chem:2010:qcp}
Liu P, Agrafiotis DK, Theobald DL.
\newblock Fast determination of the optimal rotation matrix for macromolecular
  superpositions.
\newblock J Comput Chem. 2010;31:1561--1563.

\bibitem{theobald:j-comput-chem:2011:qcp}
Liu P, Agrafiotis DK, Theobald DL.
\newblock Reply to comment on ``Fast determination of the optimal rotation
  matrix for macromolecular superpositions''.
\newblock J Comput Chem. 2011;32:185--186.

\bibitem{maccallum:prot:2011:casp-refinement}
MacCallum JL, P{\'e}rez A, Schnieders MJ, Hua L, Jacobson MP, Dill KA.
\newblock Assessment of protein structure refinement in {CASP}9.
\newblock Proteins: Structure, Function, and Bioinformatics. 2011
  Jan;79(S10):74--90.

\bibitem{zhang:curr-opin-struct-biol:2009:structure-prediction}
Zhang Y.
\newblock Protein structure prediction: when is it useful?
\newblock Current Opinion in Structural Biology. 2009 Apr;19(2):145--155.

\bibitem{raval:prot:2012:long-timescale-md-refinement}
Raval A, Piana S, Eastwood MP, Dror RO, Shaw DE.
\newblock Refinement of protein structure homology models via long, all-atom
  molecular dynamics simulations.
\newblock Proteins: Structure, Function, and Bioinformatics. 2012
  Aug;80(8):2071--2079.

\bibitem{lbfgs}
Liu DC, Nocedal J.
\newblock On the Limited Memory {BFGS} Method For Large Scale Optimization.
\newblock Mathematical Programming. 1989;45:503--528.

\bibitem{amber99sb-ildn}
Lindorff-Larsen K, anad Kim~Palmo SP, Maragakis P, Klepeis JL, Dror RO, Shaw
  DE.
\newblock Improved side-chain torsion potentials for the {Amber} {ff99SB}
  protein force field.
\newblock Proteins. 2010;78:1950--1958.

\bibitem{gbsa-obc1}
Onufriev A, Bashford D, Case DA.
\newblock Exploring protein native states and large-scale conformational
  changes with a modified generalized {Born} model.
\newblock Proteins. 2004;55(22):383--394.

\bibitem{basconi:jctc:2013:collision-rates}
Basconi JE, Shirts MR.
\newblock Effects of {Temperature} {Control} {Algorithms} on {Transport}
  {Properties} and {Kinetics} in {Molecular} {Dynamics} {Simulations}.
\newblock Journal of Chemical Theory and Computation. 2013 Jul;9(7):2887--2899.

\bibitem{tip3p}
Jorgensen WL, Chandrasekhar J, Madura JD, Impey RW, Klein ML.
\newblock Comparison of simple potential functions for simulating liquid water.
\newblock Journal of Chemical Physics. 1983;79:926--935.

\bibitem{tip4p-ew}
Horn HW, Swope WC, Pitera JW, Madura JD, Dick TJ, Hura GL, et~al.
\newblock Development of an improved four-site water model for biomolecular
  simulations: {TIP}4P-{Ew}.
\newblock The Journal of Chemical Physics. 2004 May;120(20):9665--9678.

\bibitem{molprobity:2007}
Davis IW, Leaver-Fay A, Chen VB, Block JN, Kapral GJ, Wang X, et~al.
\newblock {MolProbity}: all-atom contacts and structure validation for proteins
  and nucleic acids.
\newblock Nucleic Acids Research. 2007 Jul;35(suppl 2):W375--W383.

\bibitem{molprobity:2010}
Chen VB, Arendall WB, Headd JJ, Keedy DA, Immormino RM, Kapral GJ, et~al.
\newblock {MolProbity}: all-atom structure validation for macromolecular
  crystallography.
\newblock Acta Crystallographica Section D, Biological Crystallography. 2010
  Jan;66(Pt 1):12--21.

\bibitem{berendsen:1995:gromacs}
Berendsen HJC, van~der Spoel D, van Drunen R.
\newblock {GROMACS}: {A} message-passing parallel molecular dynamics
  implementation.
\newblock Comp Phys Comm. 1995 Sep;91(1-3):43--56.

\bibitem{lindahl:2001:gromacs}
Lindahl E, Hess B, van~der Spoel D.
\newblock {GROMACS} 3.0: a package for molecular simulation and trajectory
  analysis.
\newblock J Mol Model. 2001;7:306--317.

\bibitem{krause:nejm:2005:tks-cancer}
Krause DS, Van~Etten RA.
\newblock Tyrosine {Kinases} as {Targets} for {Cancer} {Therapy}.
\newblock New England Journal of Medicine. 2005 Jul;353(2):172--187.

\bibitem{greuber:2013:abl-cancer-review}
Greuber EK, Smith-Pearson P, Wang J, Pendergast AM.
\newblock Role of {ABL} family kinases in cancer: from leukaemia to solid
  tumours.
\newblock Nature Reviews Cancer. 2013 Aug;13(8):559--571.

\bibitem{kim:2009:src-cancer-review}
Kim LC, Song L, Haura EB.
\newblock Src kinases as therapeutic targets for cancer.
\newblock Nature Reviews Clinical Oncology. 2009 Oct;6(10):587--595.

\bibitem{liu:ncb:2006:kinase-inhibitors-inactive-conformations}
Liu Y, Gray NS.
\newblock Rational design of inhibitors that bind to inactive kinase
  conformations.
\newblock Nature Chemical Biology. 2006 Jul;2(7):358--364.

\bibitem{shukla-pande:nature-commun:2014:src-activation-msm}
Shukla D, Meng Y, Roux B, Pande VS.
\newblock Activation pathway of {Src} kinase reveals intermediate states as
  targets for drug design.
\newblock Nature Commun. 2014;5:3397.

\bibitem{chothia-lesk:embo-j:1986:rmsd-sequence-identity-relation}
Chothia C, Lesk AM.
\newblock The relation between the divergence of sequence and structure in
  proteins.
\newblock EMBO J. 1986;5:823--826.

\bibitem{roux:pnas:2013:abl-imatinib}
Lin YL, Meng Y, Jiang W, Roux B.
\newblock Explaining why {Gleevec} is a specific and potent inhibitor of {Abl}
  kinase.
\newblock Proc Natl Acad Sci USA. 2013;110:1664.

\bibitem{xu:1999:2src}
Xu W, Doshi A, Lei M, Eck MJ, Harrison SC.
\newblock Crystal {Structures} of c-{Src} {Reveal} {Features} of {Its}
  {Autoinhibitory} {Mechanism}.
\newblock Molecular Cell. 1999 May;3(5):629--638.

\bibitem{cowan-jacob:2005:1y57}
Cowan-Jacob SW, Fendrich G, Manley PW, Jahnke W, Fabbro D, Liebetanz J, et~al.
\newblock The {Crystal} {Structure} of a c-{Src} {Complex} in an {Active}
  {Conformation} {Suggests} {Possible} {Steps} in c-{Src} {Activation}.
\newblock Structure. 2005 Jun;13(6):861--871.

\bibitem{young:2006:2f4j}
Young MA, Shah NP, Chao LH, Seeliger M, Milanov ZV, Biggs WH, et~al.
\newblock Structure of the {Kinase} {Domain} of an {Imatinib}-{Resistant} {Abl}
  {Mutant} in {Complex} with the {Aurora} {Kinase} {Inhibitor} {VX}-680.
\newblock Cancer Research. 2006 Jan;66(2):1007--1014.

\bibitem{cowan-jacob:2006:2hyy}
Cowan-Jacob SW, Fendrich G, Floersheimer A, Furet P, Liebetanz J, Rummel G,
  et~al.
\newblock Structural biology contributions to the discovery of drugs to treat
  chronic myelogenous leukaemia.
\newblock Acta Crystallographica Section D: Biological Crystallography. 2006
  Dec;63(Pt 1):80--93.

\bibitem{levinson:2006:2g1t}
Levinson NM, Kuchment O, Shen K, Young MA, Koldobskiy M, Karplus M, et~al.
\newblock A {Src}-{Like} {Inactive} {Conformation} in the {Abl} {Tyrosine}
  {Kinase} {Domain}.
\newblock PLoS Biol. 2006 May;4(5):e144.

\bibitem{kannan:jmb:2005:kinase-regulation}
Kannan N, Neuwald AF.
\newblock Did {Protein} {Kinase} {Regulatory} {Mechanisms} {Evolve} {Through}
  {Elaboration} of a {Simple} {Structural} {Component}?
\newblock Journal of Molecular Biology. 2005 Sep;351(5):956--972.

\bibitem{foda:nat-comm:2015:src}
Foda ZH, Shan Y, Kim ET, Shaw DE, Seeliger MA.
\newblock A dynamically coupled allosteric network underlies binding
  cooperativity in {Src} kinase.
\newblock Nature Communications. 2015 Jan;6:5939.

\bibitem{ozkirimli:prot-sci:2008:src}
Ozkirimli E, Yadav SS, Miller WT, Post CB.
\newblock An electrostatic network and long-range regulation of {Src} kinases.
\newblock Protein Science : A Publication of the Protein Society. 2008
  Nov;17(11):1871--1880.

\bibitem{Su:2006:flt4cancer}
Su JL, Yang PC, Shih JY, Yang CY, Wei LH, Hsieh CY, et~al.
\newblock The {VEGF}-{C}/{Flt}-4 axis promotes invasion and metastasis of
  cancer cells.
\newblock Cancer Cell. 2006 Mar;9(3):209--223.

\bibitem{schnieders:biophys-j:2010:x-ray-refinement}
Fenn TD, Schnieders MJ, Brunger AT, Pande VS.
\newblock Polarizable {Atomic} {Multipole} {X}-{Ray} {Refinement}: {Hydration}
  {Geometry} and {Application} to {Macromolecules}.
\newblock Biophysical Journal. 2010 Jun;98(12):2984--2992.

\bibitem{schnieders:jctc:2011:space-group-pme}
Schnieders MJ, Fenn TD, Pande VS.
\newblock Polarizable {Atomic} {Multipole} {X}-{Ray} {Refinement}: {Particle}
  {Mesh} {Ewald} {Electrostatics} for {Macromolecular} {Crystals}.
\newblock Journal of Chemical Theory and Computation. 2011 Apr;7(4):1141--1156.

\bibitem{nagar:cell:2003:abl-1opl}
Nagar B, Hantschel O, Young MA, Scheffzek K, Veach D, Bornmann W, et~al.
\newblock Structural {Basis} for the {Autoinhibition} of c-{Abl} {Tyrosine}
  {Kinase}.
\newblock Cell. 2003 Mar;112(6):859--871.

\bibitem{shan:pnas:2009:abl}
Shan Y, Seeliger MA, Eastwood MP, Frank F, Xu H, Jensen M{\O}, et~al.
\newblock A conserved protonation-dependent switch controls drug binding in the
  {Abl} kinase.
\newblock Proceedings of the National Academy of Sciences. 2009
  Jan;106(1):139--144.

\bibitem{alexov-gunner:biophys-j:1997:mcce2}
Alexov EG, Gunner MR.
\newblock Incorporating protein conformational flexibility into the calculation
  of {pH}-dependent protein properties.
\newblock Biophys J. 1997;72:2075--2093.

\bibitem{georgescu-alexov-gunner:biophys-j:2002:mcce2}
Georgescu RE, Alexov EG, Gunner MR.
\newblock Combining conformational flexibility and continuum electrostatics for
  calculating p$K_a$s in proteins.
\newblock Biophys J. 2002;83:1731--1748.

\bibitem{song-mao-gunner:j-comput-chem:2009:mcce2}
Song Y, Mao J, Gunner MR.
\newblock {MCCE2}: Improving protein p$K_a$ calculations with extensive
  sidechain rotamer sampling.
\newblock J Comput Chem. 2009;30:2231--2247.

\bibitem{adams:prot-sci:1993:kinase-metal-ions}
Adams JA, Taylor SS.
\newblock Divalent metal ions influence catalysis and active-site accessibility
  in the {cAMP}-dependent protein kinase.
\newblock Protein Science. 1993 Dec;2(12):2177--2186.

\bibitem{sousa:book:2010:metal-ion-parameters-difficulties}
Sousa SF, Fernandes RA, Ramos MJ.
\newblock Molecular {Dynamics} {Simulations}: {Difficulties}, {Solutions} and
  {Strategies} for {Treating} {Metalloenzymes}.
\newblock In: Paneth PaDDA, editor. Kinetics and {Dynamics}: {From} {Nano}- to
  {Bio}-{Scale}. vol.~12 of Challenges and {Advances} in {Computational}
  {Chemistry} and {Physics}. Berlin: Springer Science \& Business Media; 2010.
  p. 530.

\bibitem{gaff}
Wang J, Wolf RM, Caldwell JW, Kollman PA, Case DA.
\newblock Development and testing of a general {AMBER} force field.
\newblock J Comput Chem. 2004;25:1157--1174.

\bibitem{gaff2}
Wang J, Wang W, Kollman PA, Case DA.
\newblock Automatic atom type and bond type perception in molecular mechanics
  calculations.
\newblock J Mol Graph Model. 2006;25:247260.

\end{thebibliography}



\end{document}

